`redundant_closure` false positive with `Rc<F>`/`Arc<F>`

### Summary

It seems like something changed in 1.57 that causes Clippy to suggest removing a redundant closure involving an `Arc<F>` or `Rc<F>`, but the suggestion doesn't compile, and `rustc` explicitly suggests undoing the suggestion.

### Lint Name

redundant_closure

### Reproducer

I tried this code ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6400a55b5bb84b8314059f86b20f0709)):

```rust
fn run<F: Fn() -> u64>(f: F) -> u64 {
    f()
}

fn main() {
    let f = std::sync::Arc::new(|| 5);
    let f2 = f.clone();
    
    run(move || f2());
}
```

The `redundant_closure` lint suggested rewriting the `run` call to `run(f2)`:

```
warning: redundant closure
 --> src/main.rs:9:9
  |
9 |     run(move || f2());
  |         ^^^^^^^^^^^^ help: replace the closure with the function itself: `f2`
  |
  = note: `#[warn(clippy::redundant_closure)]` on by default
``` 

But this suggestion doesn't compile, because `Arc<F>` is not a function. In fact, the compiler tells me to undo what Clippy suggested:

```
error[E0277]: expected a `Fn<()>` closure, found `Arc<[closure@src/main.rs:6:33: 6:37]>`
 --> src/main.rs:9:9
  |
9 |     run(f2);
  |     --- ^^ expected an `Fn<()>` closure, found `Arc<[closure@src/main.rs:6:33: 6:37]>`
  |     |
  |     required by a bound introduced by this call
  |
  = help: the trait `Fn<()>` is not implemented for `Arc<[closure@src/main.rs:6:33: 6:37]>`
  = note: wrap the `Arc<[closure@src/main.rs:6:33: 6:37]>` in a closure with no arguments: `|| { /* code */ }`
```

This seems to be new behavior in 1.57; the code that triggered this has been around for a while and did not trigger this lint on 1.56. The example code above also does not trigger the lint on 1.56.

### Version

```text
rustc 1.57.0 (f1edd0429 2021-11-29)
binary: rustc
commit-hash: f1edd0429582dd29cccacaf50fd134b05593bd9c
commit-date: 2021-11-29
host: aarch64-apple-darwin
release: 1.57.0
LLVM version: 13.0.0
```


### Additional Labels

@rustbot label +I-suggestion-causes-error

--- a/clippy_lints/src/eta_reduction.rs
+++ b/clippy_lints/src/eta_reduction.rs
@@ -1,6 +1,7 @@
 use clippy_utils::diagnostics::{span_lint_and_sugg, span_lint_and_then};
 use clippy_utils::higher::VecArgs;
 use clippy_utils::source::snippet_opt;
+use clippy_utils::ty::is_type_diagnostic_item;
 use clippy_utils::usage::local_used_after_expr;
 use clippy_utils::{get_enclosing_loop_or_closure, higher, path_to_local, path_to_local_id};
 use if_chain::if_chain;
@@ -12,6 +13,7 @@ use rustc_middle::ty::adjustment::{Adjust, Adjustment, AutoBorrow};
 use rustc_middle::ty::subst::Subst;
 use rustc_middle::ty::{self, ClosureKind, Ty, TypeFoldable};
 use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::symbol::sym;
 
 declare_clippy_lint! {
     /// ### What it does
@@ -113,6 +115,9 @@ impl<'tcx> LateLintPass<'tcx> for EtaReduction {
             // A type param function ref like `T::f` is not 'static, however
             // it is if cast like `T::f as fn()`. This seems like a rustc bug.
             if !substs.types().any(|t| matches!(t.kind(), ty::Param(_)));
+            let callee_ty_unadjusted = cx.typeck_results().expr_ty(callee).peel_refs();
+            if !is_type_diagnostic_item(cx, callee_ty_unadjusted, sym::Arc);
+            if !is_type_diagnostic_item(cx, callee_ty_unadjusted, sym::Rc);
             then {
                 span_lint_and_then(cx, REDUNDANT_CLOSURE, expr.span, "redundant closure", |diag| {
                     if let Some(mut snippet) = snippet_opt(cx, callee.span) {

--- a/tests/ui/eta.fixed
+++ b/tests/ui/eta.fixed
@@ -248,3 +248,14 @@ mod type_param_bound {
         take(X::fun as fn());
     }
 }
+
+// #8073 Don't replace closure with `Arc<F>` or `Rc<F>`
+fn arc_fp() {
+    let rc = std::rc::Rc::new(|| 7);
+    let arc = std::sync::Arc::new(|n| n + 1);
+    let ref_arc = &std::sync::Arc::new(|_| 5);
+
+    true.then(|| rc());
+    (0..5).map(|n| arc(n));
+    Some(4).map(|n| ref_arc(n));
+}

--- a/tests/ui/eta.rs
+++ b/tests/ui/eta.rs
@@ -248,3 +248,14 @@ mod type_param_bound {
         take(X::fun as fn());
     }
 }
+
+// #8073 Don't replace closure with `Arc<F>` or `Rc<F>`
+fn arc_fp() {
+    let rc = std::rc::Rc::new(|| 7);
+    let arc = std::sync::Arc::new(|n| n + 1);
+    let ref_arc = &std::sync::Arc::new(|_| 5);
+
+    true.then(|| rc());
+    (0..5).map(|n| arc(n));
+    Some(4).map(|n| ref_arc(n));
+}

